### ISSUE TITLE
[$500][HIGH] Clicking on an image in the email I received leads to a "Hmm... it's not here" page.

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#53770]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.73-0
**Reproducible in staging?:** Y
**Reproducible in production?:** N
**If this was caught on HybridApp, is this reproducible on New Expensify Standalone?:**
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @davidcardoza 
**Slack conversation** (hyperlinked to channel name): [#quality](https://expensify.slack.com/archives/C05LX9D6E07/p1733758949594989)

## Action Performed:

1. Send an image as user A to user B while user B is offline
2. As user B, click image in email summary

## Expected Result:
The image displayed without any issue, even on the first click
## Actual Result:
 hmm... it's not here message displayed on first click, and then the link will work after that point (on reload, or reclick).
## Workaround:
Unknown
## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Standalone
- [ ] Android: HybridApp
- [ ] Android: mWeb Chrome
- [ ] iOS: Standalone
- [ ] iOS: HybridApp
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

<details>
  <summary>Add any screenshot/video evidence</summary>

  
</details>

https://github.com/user-attachments/assets/3cd8c63b-c1fa-4af2-b903-70551fe1a429

https://github.com/user-attachments/assets/ed7822c1-3571-488a-99f5-ceb65322177e

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021882469458415172252</li>
        <li>Upwork Job ID: 1882469458415172252</li>
        <li>Last Price Increase: 2025-02-07</li>
<li>Automatic offers: </li>
<ul>
<li>NJ-2020 | Contributor | 106109491</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @allroundexperts</details>

**Comment by luacmartins** [comment]
This also happens in prod. Demoting to NAB.

**Comment by dangrous** [comment]
Hm this feels like it might be backend, but definitely open to a front end solve if we have one! Will look shortly

**Comment by dangrous** [comment]
Trying to replicate internally, apparently it's hard to get the email summaries to send.

**Comment by dangrous** [comment]
[Trying to get some help](https://expensify.slack.com/archives/C03TQ48KC/p1734019478809939) replicating this, once i can we can figure out the solution

**Comment by dangrous** [comment]
Okay so I managed to reproduce finally. The issue looks like the formation of the link from the email, vs. how you get there from the app itself. I don't yet know exactly which parts are causing the biggest issues, but here's the details.

Link from email:
```
https://dev.new.expensify.com:8082/r/3058871422043474/attachment?source=https%3A%2F%2Fwww.expensify.com.dev%2Fchat-attachments%2F5094796852905002131%2Fw_53497515bb3b313ac8e88c4efded655601d91524.png
```

URL from navigating in App:
```
https://dev.new.expensify.com:8082/attachment?source=https%3A%2F%2Fexpensify-dgalerosen.ngrok.io%2Fchat-attachments%2F5094796852905002131%2Fw_53497515bb3b313ac8e88c4efded655601d91524.png&type=r&reportID=3058871422043474&isAuthTokenRequired=true&fileName=Screenshot_2024-12-10_at_12.27.42.png
```

Differences:
```
https://dev.new.expensify.com:8082
**[EMAIL LINK HAS /r/<reportID>]**
/attachment?source=
[EMAIL HAS OLD DOT LINK, URL HAS NEW]
chat-attachments%2F5094796852905002131%2Fw_53497515bb3b313ac8e88c4efded655601d91524.png
[URL HAS &type=r&reportID=3058871422043474&isAuthTokenRequired=true&fileName=Screenshot_2024-12-10_at_12.27.42.png]
```

Playing around, I think a few things are happening:
* The report needs to be passed as a URL param, NOT as part of the url
* Possibly, the source needs to be a newdot link, though I'm not sure about that
* `fileName` and `isAuthTokenRequired` seem to possibly not be required, though that seems odd on the latter front.
* I'm not sure about `type`

The nice thing is that it's consistent, it's not about whether or not the user has access to the particular report or attachment or whatever. That is, after I successfully open it, the email link still doesn't work. So it should be easy to reproduce / update that link once I find the logic.

**Comment by dangrous** [comment]
hey @Gonals it looks like you wrote the initial URL that we use for attachment email links ([here](https://github.com/Expensify/Web-Expensify/blob/94064bacfd74563020b6c329075d56eefde895e2/lib/ReportUtils.php#L2846-L2849)). any ideas on why these aren't working anymore?

**Comment by Gonals** [comment]
That was a while ago, but it `seems` we have changed the URLs in Newdot?

**Comment by dangrous** [comment]
Yeah hrm. Okay so I've got nearly all of what we need. I've updated the backend to use the right format:

`URL_TO_NEW_DOT.'attachment?source='.$url.'&type=r&reportID='.$reportID;`

This works on dev! Yay! However, I looked into production a little bit and I'm not 100% sure it will solve the problem. I found a recent image from the social channel. The link I received in my email (which doesn't work) was
```
https://new.expensify.com/r/6549335221793525/attachment?source=https://www.expensify.com/chat-attachments/1994588430180453104/w_ab9f5e72faca59f0bc1888e27f087a52d683bf6b.jpg
```
Adapting that into what my updated backend will send, we'll get:
```
https://new.expensify.com/attachment?source=https://www.expensify.com/chat-attachments/1994588430180453104/w_ab9f5e72faca59f0bc1888e27f087a52d683bf6b.jpg&type=r&reportID=6549335221793525
```
Unfortunately, that still doesn't work. Let's keep going.

The link that actually works, via App, is
```
https://new.expensify.com/attachment?source=https%3A%2F%2Fwww.expensify.com%2Fchat-attachments%2F7162688893835784434%2Fw_c622099d5dc912677747f0b1d9b2b0d91da26598.jpg&type=r&reportID=6549335221793525&isAuthTokenRequired=true&fileName=7594B93F-3659-4022-95AF-05728EADA518.jpg
```
Adapting that into the same format from my change, it still works like this:
```
https://new.expensify.com/attachment?source=https://www.expensify.com/chat-attachments/7162688893835784434/w_c622099d5dc912677747f0b1d9b2b0d91da26598.jpg&type=r&reportID=6549335221793525
```

At this point, we're at exactly the same link [format] as my new format. But it looks like the email link still wouldn't work, even with the changes - because the reportActionID (I assume that's what the numbers after `chat-attachments` are?) and the file name (`w_....`) are different.

I'm hoping that's unrelated, and it might work if we push this change? I'm heartened by it working on dev... but that's definitely peculiar.

**Comment by dangrous** [comment]
Okay it did not fix it, I'm waiting to get a new image link so I can try to identify what else we can try.

**Comment by dangrous** [comment]
Okay it did not fix it, I'm waiting to get a new image link so I can try to identify what else we can try.

**Comment by davidcardoza** [comment]
@dangrous want me to send you an image? Don't open the DM so it will ship to your email.

**Comment by dangrous** [comment]
that would be great @davidcardoza ! I just tried sending myself one from my expensifail account but it didn't come through, even after way more than 10 minutes... maybe yours will work

**Comment by davidcardoza** [comment]
Sent a pic over via DM, it's a beaut.

**Comment by dangrous** [comment]
So in that testing ^ it DID work as expected. Do you think we should have QA re-test this one more time?

I'll also send you one @davidcardoza for another local test.

**Comment by dangrous** [comment]
Okay so it seems to work MORE of the time now, but not all the time.
* David -> Daniel worked!
* Daniel -> David worked!
* Another David -> Daniel didn't work on first click, but did on second.
* Another photo I happened to see come through worked!

I... am not sure what to do with that info haha. We're batting .750-1.000 but not certain it's enough. What do we think?

**Comment by mjasikowski** [comment]
@dangrous any updates here? Melvin is getting anxious

**Comment by dangrous** [comment]
oops hi, was on break for the holidays. In further testing, this still doesn't work reliably. I've got some more images that were sent in social, so I'm going to try to investigate them further and see what I can do

**Comment by dangrous** [comment]
Okay so I noticed a couple things:
* All image links in emails direct to new.expensify.com. However, in the URL query the `source` param of the image is either `www.expensify.com` or `staging.expensify.com`. My guess is this is due to which platform the user uploaded the image on. **However, if these don't match, the link doesn't work**. prod and prod works, stag and stag works, but not a mix. So **we should probably force the source link to be prod**.
* Weirdly, when you click on a link like `https://new.expensify.com/attachment?source=https://www.expensify.com/chat-attachments/4324233214072273054/w_786e7a0caedb4e522f56c6b328a1f1b8f88be853.jpg&type=r&reportID=6549335221793525` , when the URL resolves in the browser the source is `https:/www.expensify...` - it loses one of the slashes. I haven't figured out why this is yet. It happens on both staging and prod urls.
* The `source` of in-app links are always URL encoded, but that seems separated from whether or not the links work. decoded good links are still good, encoded bad links are still bad.
* The links clicked in app always have `&isAuthTokenRequired=true&fileName=[whatever].png` at the end but I'm pretty sure that it's not really necessary? it's hard to tell since there's so much else going on.
* Lastly, it's so weird that it sometimes works now. It seems to be more often NOT in #social, but that doesn't really help. Still, sometimes is better than none of the time?

Sooo.. not really closer but at least we have some threads to pull on? I should probably try to figure out on the other side what happens - like when the app gets a deep link to an image, how does it parse it and what is it looking for. Rather than just looking to make them match exactly.

Side note: do we care that email links are always prod? What if you're logged in on staging and not prod? I guess it's mostly an internal question.

**Comment by dangrous** [comment]
So good news - I've figured out how to get the file token and the authtoken required in the attachment link; bad news is it doesn't help. Asking for some help in Slack to point me towards where we handle these deeplinks in App so I can see if there's anything I can do on that side.

**Comment by dangrous** [comment]
Okay I _think_ I've got a PR up that will fix this once and for all. Uncovered a separate bug with videos but that doesn't have to be dealt with here.

**Comment by dangrous** [comment]
PR is on staging. Fingers crossed this works when we get to prod!

**Comment by dangrous** [comment]
Okay this is on prod after a little hiccup with a deploy blocker. Let's see if it works out! I'll keep an eye out for images.

**Comment by dangrous** [comment]
hey @davidcardoza can you send me another image when you get a sec? Either in a room or in a DM (or maybe both?)

**Comment by dangrous** [comment]
just attempted to send myself an image on prod and one on staging, those should come through shortly and give us a little more insight? maybe?

**Comment by dangrous** [comment]
weird those didn't come through at all... Didn't see any errors though in the logs so not sure why

**Comment by dangrous** [comment]
Okay got some images in emails. So we've solved one problem here - the links now seem to be in the correct format.

However!

That annoyingly doesn't mean that when you click on one, you're able to see the image. It seems like it does not work on the first click, but then if you click again (or refresh the page), you will see it.

I think this is because on the first click, you don't actually have the report action loaded in your Onyx yet. Like I think if it's not there right away, it just immediately gives up and thinks the content doesn't exist. So I need to figure out what happens when we go to a direct report action link. I assume it triggers an OpenReport, but something is causing that image viewer not to wait for that call to finish.

We're definitely closer, though!

**Comment by dangrous** [comment]
Asking for some assistance in Slack. https://expensify.slack.com/archives/C01GTK53T8Q/p1737491229973979

**Comment by dangrous** [comment]
hey @allroundexperts - there's a lot of backstory on this issue, but we think we've solved the backend issue and now need someone to investigate the frontend and see if there's a fix there.

**Comment by dangrous** [comment]
Open to external, if you want to be the C+ - but you're also welcome to grab it if you'd prefer that

**Comment by allroundexperts** [comment]
Let's see if we can get some proposals first!

**Comment by dangrous** [comment]
Waiting for proposals. I'm tempted to make this weekly but it's a pretty annoying bug so would be good to get some action soon

**Comment by dangrous** [comment]
Melvin, not overdue

**Comment by dangrous** [comment]
I'm gonna bump the price on this one!

**Comment by jaydamani** [comment]
I think I found the root cause but waiting for an email to test. 
From my investigation so far:
Navigating to `/attachments` loads [ReportAttachments](https://github.com/Expensify/App/blob/d2bccd32f2ef36b44359ee6be1e06e73ff9f9149/src/pages/home/report/ReportAttachments.tsx#L16) which based on [this](https://github.com/Expensify/App/blob/d2bccd32f2ef36b44359ee6be1e06e73ff9f9149/src/pages/home/report/ReportAttachments.tsx#L59) will show the not found page when report is not loaded and the app stops loading. 

So, for our case if the report is not already loaded or is not loaded by openApp, it will show the not found page. To solve this, we can send an openAttachment (new api that can be alias of openReport) request if the report is not loaded when user opens `ReportAttachments`.
Waiting for an email to be able to test properly before making a proposal.

Also, again untested but the filename param is used for when [downloading](https://github.com/Expensify/App/blob/d2bccd32f2ef36b44359ee6be1e06e73ff9f9149/src/components/AttachmentModal.tsx#L253) the file
cc: @dangrous @allroundexperts for thoughts

**Comment by jaydamani** [comment]
So, I was able to replicate it only once then the onyx data for report was added  by openApp. So, this might be the right path but not sure about when is a report not loaded by openApp. Hoping if someone can point that out :D

**Comment by dangrous** [comment]
yeah unfortunately it's annoying to test because you only get that one shot per image haha

**Comment by mjasikowski** [comment]
@allroundexperts bump for review.

**Comment by allroundexperts** [comment]
@NJ-2020 Can you explain why it starts to work when you reload or click on the image again?

**Comment by allroundexperts** [comment]
> [@allroundexperts](https://github.com/allroundexperts) The bug still reproducible even we click/reload the image
> 
>  Screen.Recording.2025-02-12.at.06.02.09.mov

Thanks for the quick turnaround. @NJ-2020's [proposal](https://github.com/Expensify/App/issues/53770#issuecomment-2636185457) has the correct RCA and the proposed solution fixes the problem. Let's go with them.

🎀 👀 🎀 C+ reviewed

**Comment by dangrous** [comment]
how are we looking on this one @allroundexperts @NJ-2020 ?

**Comment by allroundexperts** [comment]
Sorry, I was travelling last week. I am back at my workplace, and will review this today.

**Comment by dangrous** [comment]
how are we looking on this one? Looks like we found some additional complexity while working on the PR?

**Comment by dangrous** [comment]
@allroundexperts @NJ-2020 do y'all need any help on the PR? What's our status?

**Comment by JmillsExpensify** [comment]
PR in active review.

**Comment by dangrous** [comment]
This fix is on staging!

**Comment by allroundexperts** [comment]
Hey @dangrous!

Can we review the compensation for this issue again? This was very tricky, and it took us a lot of back and forth (~4 months) to fix all the edge cases. I think it should be doubled. Let me know what you think.

Thanks!

**Comment by dangrous** [comment]
Yeah I think that's totally fair - @JmillsExpensify let's increase this one to 1k since it took a lot of effort, and was particularly annoying too haha

**Comment by allroundexperts** [comment]
# BugZero Checklist:

- [x] **[Contributor]** Classify the bug:

<details>
<summary>Bug classification</summary>


Source of bug:
  - [x] 1a. Result of the original design (eg. a case wasn't considered)
  - [ ] 1b. Mistake during implementation
  - [ ] 1c. Backend bug
  - [ ] 1z. Other:

Where bug was reported:
  - [ ] 2a. Reported on production (eg. bug slipped through the normal regression and PR testing process on staging)
  - [ ] 2b. Reported on staging (eg. found during regression or PR testing)
  - [ ] 2d. Reported on a PR
  - [x] 2z. Other: Reported on staging, but was on production as well.

Who reported the bug:
  - [ ] 3a. Expensify user
  - [x] 3b. Expensify employee
  - [ ] 3c. Contributor
  - [ ] 3d. QA
  - [ ] 3z. Other:

</details>

- [x] **[Contributor]** The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake.

    Link to comment: There isn't any PR which caused this. It seems like this case was never considered.

- [x] **[Contributor]** If the regression was CRITICAL (e.g. interrupts a core flow) A discussion in [#expensify-open-source](https://app.slack.com/client/E047TPA624F/C01GTK53T8Q) has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner.

    Link to discussion: N/A

- [x] **[Contributor]** If it was decided to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps using the template below to ensure the same bug will not reach production again.

- [ ] **[BugZero Assignee]** Create a GH issue for creating/updating the regression test once above steps have been agreed upon.

    Link to issue:

## Regression Test Proposal
### Precondition:


- N/A

### Test:


1. Login to two different accounts as user A and user B
2. Send an image as user A to user B while user B is offline
3. Wait for the message email to reach user B
4. As user B, click image in email summary

Verify that the image is displayed in the app without any issue

Do we agree 👍 or 👎

**Comment by NJ-2020** [comment]
Bump @JmillsExpensify for payment

**Comment by JmillsExpensify** [comment]
Payment summary:

 - $1,000 for @NJ-2020 via Upwork
 - $1,000 for @allroundexperts via NewDot

**Comment by allroundexperts** [comment]
@JmillsExpensify The bounty for this is $1000 as per https://github.com/Expensify/App/issues/53770#issuecomment-2931217000

**Comment by JmillsExpensify** [comment]
Updated

**Comment by JmillsExpensify** [comment]
Contributor paid out and QA tests added.

**Comment by garrettmknight** [comment]
Payment summary:

- $1,000 for @NJ-2020 via Upwork
- $1,000 for @allroundexperts via NewDot
